### PR TITLE
Fix call `engine.destroy()` crash  in `Script`

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Main modify:
- Use delay destroy rule to ignore crash bug
- Optimization Script logic.

### CODE REVIEW

File: packages/core/src/Script.ts [hunk 1]
```diff
@@ -219,20 +219,11 @@ export class Script extends Component {
...
-    }
-    if (this._onPhysicsUpdateIndex !== -1) {
-      componentsManager.removeOnPhysicsUpdateScript(this);
-    }
-    if (this._entityScriptsIndex !== -1) {
-      this._entity._removeScript(this);
-    }
+    componentsManager.removeOnUpdateScript(this);
+    componentsManager.removeOnLateUpdateScript(this);
+    componentsManager.removeOnPhysicsUpdateScript(this);
```
Comment (singlecoder) [inline packages/core/src/Script.ts:224]: You don't need to consider the case of -1, is it guaranteed not to have?

File: packages/core/src/Engine.ts [hunk 2]
```diff
@@ -300,10 +309,15 @@ export class Engine extends EventDispatcher {
...
-      componentsManager.handlingInvalidScripts();
     }
 
     engineFeatureManager.callFeatureMethod(this, "postTick", [this, this._sceneManager._activeScene]);
+
+    // Engine is complete delayed destruction mechanism
+    if (this._waittingDestroy) {
+      this._destroy();
+    }
+    componentsManager.handlingInvalidScripts();
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/Engine.ts:320]: Why not put it before destroy?

File: packages/core/src/Script.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/Script.ts:224]: After this PR https://github.com/oasis-engine/engine/pull/1055 fixed, call onDisable before onEnable maybe never happen.

File: packages/core/src/Engine.ts (hunk 2 above)
Comment (GuoLei1990) [reply packages/core/src/Engine.ts:320]: That not wor,  `this._sceneManager._destroy();`  is in `this._destroy()`

Comment (cptbtptpbcptdtptp) [reply packages/core/src/Engine.ts:320]: For example:
```
onDestroy(): void {
     console.log('Destroyed at', this.engine.time.nowTime());
}
```

File: packages/core/src/Script.ts (hunk 1 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/Script.ts:224]: > After this PR #1055 fixed, call onDisable before onEnable maybe never happen.

Even if repaired, the subscript of `onUpdate` may be -1.
```
if (this.onUpdate !== prototype.onUpdate) {
        componentsManager.addOnUpdateScript(this);
}
```

File: packages/core/src/Engine.ts (hunk 2 above)
Comment (cptbtptpbcptdtptp) [reply packages/core/src/Engine.ts:320]: Understood.

Comment (GuoLei1990) [reply packages/core/src/Engine.ts:320]: done

File: packages/core/src/Script.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/Script.ts]: done

**Review by cptbtptpbcptdtptp — APPROVED** [review]
+1

**Review by singlecoder — APPROVED** [review]
+1

**Pull request merged**